ext/json/ext/parser/parser.c

```diff
@@ -767,6 +767,32 @@ static VALUE json_string_unescape(JSON_ParserState *state, const char *string, c
 
 #define MAX_FAST_INTEGER_SIZE 18
 
+#if defined(__BYTE_ORDER__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+// From: https://lemire.me/blog/2022/01/21/swar-explained-parsing-eight-digits/
+// Additional References:
+// https://johnnylee-sde.github.io/Fast-numeric-string-to-int/
+// http://0x80.pl/notesen/2014-10-12-parsing-decimal-numbers-part-1-swar.html
+static inline uint32_t parse_eight_digits_unrolled(const char *p) {
+    uint64_t val;
+    memcpy(&val, p, sizeof(uint64_t));
+    const uint64_t mask = 0x000000FF000000FF;
+    const uint64_t mul1 = 0x000F424000000064; // 100 + (1000000ULL << 32)
+    const uint64_t mul2 = 0x0000271000000001; // 1 + (10000ULL << 32)
+    val -= 0x3030303030303030;
+    val = (val * 10) + (val >> 8); // val = (val * 2561) >> 8;
+    val = (((val & mask) * mul1) + (((val >> 16) & mask) * mul2)) >> 32;
+    return (uint32_t) val;
+}
+
+// From: https://github.com/simdjson/simdjson/blob/32b301893c13d058095a07d9868edaaa42ee07aa/include/simdjson/generic/numberparsing.h#L333
+// Branchless version of: http://0x80.pl/articles/swar-digits-validate.html
+static inline int has_eight_consecutive_digits(const char *p) {
+    uint64_t val;
+    memcpy(&val, p, sizeof(uint64_t));
+    return (((val & 0xF0F0F0F0F0F0F0F0) | (((val + 0x0606060606060606) & 0xF0F0F0F0F0F0F0F0) >> 4)) == 0x3333333333333333);
```

Comment:
I'm thinking we could use that trick combined with `clz` (or similar) to know how many consecutive digits we have.

I suspect 8 consecutive digits aren't that common, but if we also had a 4 digits (`uint32_t`) version and a fast dispatch, that could help on more benchmarks.

Reply:
Actually, I think we can simply do `(comp & 0xFFFFFFFF) == 0x33333333` to check for 4 consecutive digits.

Reply:
4bytes version:

```c
static inline uint32_t parse_four_digits_unrolled(const char *p) {
    uint64_t large_val;
    memcpy(&large_val, p, sizeof(uint64_t));
    uint32_t val = (uint32_t)large_val;

    const uint32_t mask = 0x000000FF;
    const uint32_t mul1 = 100;
    val -= 0x30303030;
    val = (val * 10) + (val >> 8); // val = (val * 2561) >> 8;
    val = ((val & mask) * mul1) + (((val >> 16) & mask));
    return (uint32_t)val;
}
```